Introduce documentation for using GameLift Anywhere (#402)
o3de/o3de-multiplayersample: MPSGameLift/Documentation/GameLift.md

Author: I originally went for a separate file as you suggest, but over 50% of the steps would have been duplicated. So I would've either had to ask the reader to reference the other file and go back and forth, or copy a good chunk of the content and made it harder to maintain.

Honestly I'm fine with either option. If we want to keep all in one doc, I will definitely add a bit more of an intro to explain the differences.

# MultiplayerSample Project with Amazon GameLift

> This README covers optional setup, testing and running on [Amazon GameLift](https://aws.amazon.com/gamelift/), an AWS service to make hosting and scaling game servers easier. It also provides guidance on how to test the Amazon GameLift integration on your local machine via GameLift Anywhere.

Reviewer: ```suggestion
This README covers optional setup, testing and running on [Amazon GameLift](https://aws.amazon.com/gamelift/), an AWS service to make hosting and scaling game servers easier. It also provides guidance on how to test the Amazon GameLift integration on your local machine via [Amazon GameLift Anywhere](https://docs.aws.amazon.com/gamelift/latest/developerguide/fleets-creating-anywhere.html).
```

    ---
    

> It's important to make sure that the bootstrap.game.profile.setreg file has been added to one of the seed files. (also add debug if you want to support debug builds)

Reviewer: ```suggestion
It's important to make sure that the _bootstrap.game.profile.setreg_ file has been added to one of the seed files. (also add debug if you want to support debug builds)
```

---
**NOTE**

This will not build the game client. If you need the client for testing, also run this command:

> ```

Reviewer: ```suggestion
```sh
```
to improve syntax highlighting

    ---
It's important to make sure that the bootstrap.game.profile.setreg file has been added to one of the seed files. (also add debug if you want to support debug builds)

---
**NOTE**

This will not build the game client. If you need the client for testing, also run this command:
```

> cmake --build build\windows_mono --target MultiplayerSample.GameLauncher --config profile -- /m /nologo

Reviewer 2: I would add context that the export_gamelift_server_package sets up to build things monolithic etc

Also could we add this as an option ie -build-test-client = true? to that script so it just does it?

Author: Yes, we discussed having an option with Gene. If we want to do that, we can spin up a separate task for it and either of us can pick it up.

    ```

## Test your setup locally with GameLift Anywhere

Before you set up a real GameLift fleet, you can use GameLift Anywhere to quickly verify your build.
This guide mirrors the steps on the [Amazon GameLift Anywhere Developer Guide](https://docs.aws.amazon.com/gamelift/latest/developerguide/fleets-creating-anywhere.html)

### Create a Location

Create a location for your custom resources. This is a categorization akin to AWS regions.
When creating your custom location, the location name must start with `custom-`.

```sh
aws gamelift create-location --location-name custom-location-1 --region <Region>
```

If the operation was successful, the console will display the JSON result.
Record the `LocationName` for the next step. Example: custom-location-1

### Create a fleet

Creating an Anywhere fleet is a much faster process compared to creating a regular AWS fleet, which usually takes about an hour to setup.

```sh
aws gamelift create-fleet --name AnywhereFleet --compute-type ANYWHERE --locations Location=custom-location-1 --region <Region>
```

If the operation was successful, the console will display the JSON result.

> Record the `FleetId` for the next steps. Example: fleet-1a23bc4d-456e-78fg-h9i0-jk1l23456789

Reviewer: ```suggestion
Record the `FleetId` for the next steps. Example: **fleet-1a23bc4d-456e-78fg-h9i0-jk1l23456789**
```

> ### Register your local machina as a Compute

Reviewer: ```suggestion
### Register your local machine as a Compute
```

Register your local machine as a GameLift Anywhere Compute.
For ease of testing, we assume the Server and Client will be run on the same machine; so we can pass localhost (`127.0.0.1`) as the IP address.
If your machine is accessible via a public IP address, change that value as appropriate.

```sh
aws gamelift register-compute --compute-name CustomCompute1 --fleet-id <FleetId> --ip-address 127.0.0.1 --location custom-location-1 --region <Region>
```

If the operation was successful, the console will display the JSON result.

> Record the `ComputeName` for the next steps. Example: CustomCompute1

Reviewer: ```suggestion
Record the `ComputeName` for the next steps. Example: **CustomCompute1**
```

> Also record the `GameLiftServiceSdkEndpoint` for later. Example: wss://us-west-2.api.amazongamelift.com

Reviewer: ```suggestion
Also record the `GameLiftServiceSdkEndpoint` for later. Example: **wss://us-west-2.api.amazongamelift.com**
```

### Get Compute auth token

```sh
aws gamelift get-compute-auth-token --fleet-id <FleedId> --compute-name <ComputeName>
```

If the operation was successful, the console will display the JSON result.

> Record the `AuthToken` for the next steps. Example: 123a4b5c-d6e7-8fgh-9i01-2jklm34no567

Reviewer: ```suggestion
Record the `AuthToken` for the next steps. Example: **123a4b5c-d6e7-8fgh-9i01-2jklm34no567**
```

## Start an instance of the Game Server executable on your machine

To ensure GameLift recognizes your local machine as a Compute that is available to start game sessions, start a server locally with the appropriate credentials.

Notes:
- In the `HostId` property should be filled with the `ComputeName`
- `ProcessId` can be omitted; a unique default `ProcessId` will be generated out of the timestamp.

```sh
C:\GameLiftPackageWindows\MultiplayerSample.ServerLauncher.exe --rhi=null -NullRenderer -bg_ConnectToAssetProcessor=0 --console-command-file=launch_server.cfg --sv_dedicated_host_onstartup=false --sv_gameLiftEnabled=true --sv_gameliftAnywhereEnabled=true --sv_gameliftAnywhereWebSocketUrl="<WebSocketUrl>" --sv_gameliftAnywhereAuthToken="<AuthToken>" --sv_gameliftAnywhereFleetId="<FleetId>" --sv_gameliftAnywhereHostId="<ComputeName>" --sv_gameliftAnywhereProcessId="<ProcessId>"
```

### Create a Game Session

```sh
aws gamelift create-game-session --region <Region> --location custom-location-1 --fleet-id <FleetId> --name GameSession1 --maximum-player-session-count 3
```

If the operation was successful, the console will display the JSON result.

> Record the `GameSessionId` for the next steps. Example: arn:aws:gamelift:us-west-2::gamesession/fleet-1a23bc4d-456e-78fg-h9i0-jk1l23456789/custom-location-1/gsess-ab1cd2ef-3gh4-5678-ijk9-0l1mn2o345p6

Reviewer: ```suggestion
Record the `GameSessionId` for the next steps. Example: **arn:aws:gamelift:us-west-2::gamesession/fleet-1a23bc4d-456e-78fg-h9i0-jk1l23456789/custom-location-1/gsess-ab1cd2ef-3gh4-5678-ijk9-0l1mn2o345p6**
```

If the operation fails, make sure the server is running. Ensure that `InitSDK` and `ProcessReady` calls were successful.

### Start Client

```sh
.\build\windows\bin\profile\MultiplayerSample.GameLauncher.exe -bg_ConnectToAssetProcessor=0 --loadlevel="mpsgamelift/prefabs/GameLiftConnectJsonMenu.spawnable"
```

Once started, the client should show a text area where the session information needs to be pasted into. You may need to press `~` on your keyboard to open the console and release the cursor from being bound to the client window.
It is recommended to prepare this JSON object in advance as the `PlayerSessionId` generated in the next step is only valid for 60 seconds.

```json
{ "GameSessionId": "<GameSessionId>", "PlayerId": "PlayerId", "PlayerSessionId": "<PlayerSessionId>" }
```

### Create a Player Session

Note: `PlayerId` from the JSON above and `--player-id` in the command below do not need to be the same

```sh
aws gamelift create-player-session --region <Region> --game-session-id <GameSessionId> --player-id Player1
```

If the operation was successful, the console will display the JSON result.

> Record the `PlayerSessionId` for the next steps. Example: psess-1a2b3c45-d6e7-89fg-0hij-12kl34m56no7

Reviewer: ```suggestion
Record the `PlayerSessionId` for the next steps. Example: **psess-1a2b3c45-d6e7-89fg-0hij-12kl34m56no7**
```

Record the `AuthToken` for the next steps. Example: 123a4b5c-d6e7-8fgh-9i01-2jklm34no567

## Start an instance of the Game Server executable on your machine

To ensure GameLift recognizes your local machine as a Compute that is available to start game sessions, start a server locally with the appropriate credentials.

Notes:
- In the `HostId` property should be filled with the `ComputeName`
- `ProcessId` can be omitted; a unique default `ProcessId` will be generated out of the timestamp.

```sh
C:\GameLiftPackageWindows\MultiplayerSample.ServerLauncher.exe --rhi=null -NullRenderer -bg_ConnectToAssetProcessor=0 --console-command-file=launch_server.cfg --sv_dedicated_host_onstartup=false --sv_gameLiftEnabled=true --sv_gameliftAnywhereEnabled=true --sv_gameliftAnywhereWebSocketUrl="<WebSocketUrl>" --sv_gameliftAnywhereAuthToken="<AuthToken>" --sv_gameliftAnywhereFleetId="<FleetId>" --sv_gameliftAnywhereHostId="<ComputeName>" --sv_gameliftAnywhereProcessId="<ProcessId>"
```

### Create a Game Session

```sh
aws gamelift create-game-session --region <Region> --location custom-location-1 --fleet-id <FleetId> --name GameSession1 --maximum-player-session-count 3
```

> If the operation was successful, the console will display the json result.

Reviewer 2: ```suggestion
If the operation was successful, the console will display the JSON result.
```

Author: Fixed everywhere!

Record the `GameSessionId` for the next steps. Example: arn:aws:gamelift:us-west-2::gamesession/fleet-1a23bc4d-456e-78fg-h9i0-jk1l23456789/custom-location-1/gsess-ab1cd2ef-3gh4-5678-ijk9-0l1mn2o345p6

> If the operation fails, make sure the server is running and the `InitSDK` call was successful.

Reviewer 2: ```suggestion
If the operation fails, make sure the server is running. Ensure that `InitSDK` and `ProcessReady` calls were successful.
```

```sh
aws gamelift upload-build --operating-system WINDOWS_2016 --build-root C:\GameLiftPackageWindows\ --name MultiplayerSample --build-version v1.0 --region us-west-2
```

> Record BuildId for the next step. Example: build-1a23bc4d-456e-78fg-h9i0-jk1l23456789

Reviewer: ```suggestion
Record BuildId for the next step. Example: **build-1a23bc4d-456e-78fg-h9i0-jk1l23456789**
```

---

> Record the FleetId for the next step. Example: fleet-1a23bc4d-456e-78fg-h9i0-jk1l23456789

Reviewer: ```suggestion
Record the FleetId for the next step. Example: **fleet-1a23bc4d-456e-78fg-h9i0-jk1l23456789**
```
